File: src-tauri/src/state/user_profile.rs
use serde::{Deserialize, Serialize};
use std::collections::HashMap;

/// 热词来源
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "lowercase")]
pub enum HotWordSource {
    /// 用户手动添加
    User,
    /// 从纠错中自动学习
    Learned,
}

/// 纠错来源
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Default)]
#[serde(rename_all = "lowercase")]
pub enum CorrectionSource {
    /// AI 润色自动学习
    #[default]
    Ai,
    /// 用户手动纠错
    User,
}

/// 热词条目
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct HotWord {
    pub text: String,
    /// 权重 1-5，越高越优先
    pub weight: u8,
    pub source: HotWordSource,
    pub use_count: u32,
    /// Unix 时间戳（秒）
    pub last_used: u64,
}

/// 纠错模式记录
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct CorrectionPattern {
    /// ASR 原始输出片段
    pub original: String,
    /// 纠正后的片段
    pub corrected: String,
    /// 出现次数
    pub count: u32,
    /// 最近一次出现的时间戳
    pub last_seen: u64,
    /// 纠错来源
    #[serde(default)]
    pub source: CorrectionSource,
}

/// 词频条目
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct VocabEntry {
    pub count: u32,
    pub last_seen: u64,
}

/// 用户画像
#[derive(Debug, Clone, Serialize, Deserialize, Default)]
pub struct UserProfile {
    pub hot_words: Vec<HotWord>,
    pub correction_patterns: Vec<CorrectionPattern>,
    pub vocab_frequency: HashMap<String, VocabEntry>,
    pub total_transcriptions: u64,
    /// 上次更新时间（Unix 时间戳）
    pub last_updated: u64,
    /// LLM 后端配置
    pub llm_provider: LlmProviderConfig,
    /// 翻译目标语言（None = 关闭翻译，非空 = 开启并翻译为该语言）
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub translation_target: Option<String>,
    /// 翻译模式独立热键
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub translation_hotkey: Option<String>,
    /// 用户自定义润色指令
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub custom_prompt: Option<String>,
    /// 助手模式独立热键
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub assistant_hotkey: Option<String>,
    /// 助手模式附加系统提示词
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub assistant_system_prompt: Option<String>,
    /// 助手模式是否附带当前屏幕截图作为上下文
    #[serde(default)]
    pub assistant_screen_context_enabled: bool,
    /// 用户手动删除后，不再自动学习回来的热词黑名单
    #[serde(default, skip_serializing_if = "Vec::is_empty")]
    pub blocked_hot_words: Vec<String>,
}

/// API 协议格式
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Default)]
#[serde(rename_all = "snake_case")]
pub enum ApiFormat {
    #[default]
    OpenaiCompat,
    Anthropic,
}

/// 推理/思考模式（跨供应商抽象层）
#[derive(Debug, Clone, Copy, Serialize, Deserialize, PartialEq, Eq, Default)]
#[serde(rename_all = "snake_case")]
pub enum LlmReasoningMode {
    /// 不下发任何推理参数，走供应商默认行为
    #[default]
    ProviderDefault,
    /// 尽量关闭或压低思考
    Off,
    /// 偏轻量
    Light,
    /// 标准
    Balanced,
    /// 偏深度
    Deep,
}

/// 用户自定义的 LLM 服务商
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct CustomProvider {
    pub id: String,
    pub name: String,
    pub base_url: String,
    pub model: String,
    #[serde(default)]
    pub api_format: ApiFormat,
}

/// LLM 后端配置
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct LlmProviderConfig {
    /// 当前使用的后端: 预置 key 或 custom provider id
    pub active: String,
    /// 旧字段，迁移兼容
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub custom_base_url: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub custom_model: Option<String>,
    /// 跨供应商的思考模式抽象层
    #[serde(default)]
    pub reasoning_mode: LlmReasoningMode,
    /// AI 润色链路的独立思考模式；为空时回退到旧的 reasoning_mode
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub polish_reasoning_mode: Option<LlmReasoningMode>,
    /// AI 助手链路的独立思考模式；为空时回退到旧的 reasoning_mode
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub assistant_reasoning_mode: Option<LlmReasoningMode>,
    /// AI 助手是否使用不同于润色的独立模型
    #[serde(default)]
    pub assistant_use_separate_model: bool,
    /// AI 助手独立模型；仅在 assistant_use_separate_model = true 时生效
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub assistant_model: Option<String>,
    /// 用户自定义服务商列表
    #[serde(default, skip_serializing_if = "Vec::is_empty")]
    pub custom_providers: Vec<CustomProvider>,
}

impl Default for LlmProviderConfig {
    fn default() -> Self {
        Self {
            active: "cerebras".to_string(),
            custom_base_url: None,
            custom_model: None,
            reasoning_mode: LlmReasoningMode::ProviderDefault,
            polish_reasoning_mode: None,
            assistant_reasoning_mode: None,
            assistant_use_separate_model: false,
            assistant_model: None,
            custom_providers: Vec::new(),
        }
    }
}

impl LlmProviderConfig {
    fn is_builtin_provider(provider: &str) -> bool {
        matches!(
            provider,
            "cerebras" | "openai" | "deepseek" | "siliconflow" | "custom"
        )
    }

    pub fn resolve_active_provider(&self) -> String {
        if Self::is_builtin_provider(&self.active)
            || self.custom_providers.iter().any(|p| p.id == self.active)
        {
            return self.active.clone();
        }

        self.custom_providers
            .last()
            .map(|provider| provider.id.clone())
            .unwrap_or_else(|| "cerebras".to_string())
    }

    pub fn fallback_provider_after_removal(&self, removed_id: &str) -> String {
        if self.active != removed_id {
            return self.resolve_active_provider();
        }

        if let Some(index) = self
            .custom_providers
            .iter()
            .position(|p| p.id == removed_id)
        {
            if index > 0 {
                return self.custom_providers[index - 1].id.clone();
            }

            if let Some(next) = self
                .custom_providers
                .iter()
                .enumerate()
                .rev()
                .find(|(idx, _)| *idx != index)
                .map(|(_, provider)| provider.id.clone())
            {
                return next;
            }
        }

        "cerebras".to_string()
    }

    pub fn polish_reasoning_mode(&self) -> LlmReasoningMode {
        self.polish_reasoning_mode.unwrap_or(self.reasoning_mode)
    }

    pub fn assistant_reasoning_mode(&self) -> LlmReasoningMode {
        self.assistant_reasoning_mode.unwrap_or(self.reasoning_mode)
    }

    pub fn assistant_model(&self) -> Option<&str> {
        if !self.assistant_use_separate_model {
            return None;
        }
        self.assistant_model
            .as_deref()
            .map(str::trim)
            .filter(|value| !value.is_empty())
    }
}

impl UserProfile {
    /// 获取按权重排序的热词文本列表（用于 ASR 注入）
    pub fn get_hot_word_texts(&self, limit: usize) -> Vec<String> {
        let mut words: Vec<&HotWord> = self.hot_words.iter().collect();
        words.sort_by(|a, b| b.weight.cmp(&a.weight).then(b.use_count.cmp(&a.use_count)));
        words
            .into_iter()
            .take(limit)
            .map(|w| w.text.clone())
            .collect()
    }

    /// 根据输入文本动态检索相关纠错模式
    ///
    /// 策略：精确子串匹配优先（input 中包含 pattern.original），
    /// 不足 limit 条时用高频纠错兜底。User 来源始终优先于 Ai。
    pub fn get_relevant_corrections(&self, input: &str, limit: usize) -> Vec<&CorrectionPattern> {
        let source_ord = |s: &CorrectionSource| match s {
            CorrectionSource::User => 0,
            CorrectionSource::Ai => 1,
        };

        // 第一轮：精确子串命中
        let mut matched: Vec<&CorrectionPattern> = self
            .correction_patterns
            .iter()
            .filter(|p| !p.original.is_empty() && input.contains(&p.original))
            .collect();
        matched.sort_by(|a, b| {
            source_ord(&a.source)
                .cmp(&source_ord(&b.source))
                .then(b.count.cmp(&a.count))
        });
        matched.truncate(limit);

        if matched.len() >= limit {
            return matched;
        }

        // 第二轮：高频兜底，补足剩余名额
        let remaining = limit - matched.len();
        let matched_set: std::collections::HashSet<(&str, &str)> = matched
            .iter()
            .map(|p| (p.original.as_str(), p.corrected.as_str()))
            .collect();

        let mut fallback: Vec<&CorrectionPattern> = self
            .correction_patterns
            .iter()
            .filter(|p| !matched_set.contains(&(p.original.as_str(), p.corrected.as_str())))
            .collect();
        fallback.sort_by(|a, b| {
            source_ord(&a.source)
                .cmp(&source_ord(&b.source))
                .then(b.count.cmp(&a.count))
        });

        matched.extend(fallback.into_iter().take(remaining));
        matched
    }
}

#[cfg(test)]
mod tests {
    use super::{ApiFormat, CustomProvider, LlmProviderConfig, LlmReasoningMode};

    fn custom_provider(id: &str) -> CustomProvider {
        CustomProvider {
            id: id.to_string(),
            name: id.to_string(),
            base_url: format!("https://{id}.example.com"),
            model: format!("model-{id}"),
            api_format: ApiFormat::OpenaiCompat,
        }
    }

    #[test]
    fn resolves_invalid_active_provider_to_latest_custom_provider() {
        let config = LlmProviderConfig {
            active: "missing".to_string(),
            custom_base_url: None,
            custom_model: None,
            reasoning_mode: Default::default(),
            polish_reasoning_mode: None,
            assistant_reasoning_mode: None,
            assistant_use_separate_model: false,
            assistant_model: None,
            custom_providers: vec![custom_provider("a"), custom_provider("b")],
        };

        assert_eq!(config.resolve_active_provider(), "b");
    }

    #[test]
    fn falls_back_to_previous_provider_after_removal() {
        let config = LlmProviderConfig {
            active: "b".to_string(),
            custom_base_url: None,
            custom_model: None,
            reasoning_mode: Default::default(),
            polish_reasoning_mode: None,
            assistant_reasoning_mode: None,
            assistant_use_separate_model: false,
            assistant_model: None,
            custom_providers: vec![
                custom_provider("a"),
                custom_provider("b"),
                custom_provider("c"),
            ],
        };

        assert_eq!(config.fallback_provider_after_removal("b"), "a");
    }

    #[test]
    fn falls_back_to_last_remaining_provider_when_removing_first() {
        let config = LlmProviderConfig {
            active: "a".to_string(),
            custom_base_url: None,
            custom_model: None,
            reasoning_mode: Default::default(),
            polish_reasoning_mode: None,
            assistant_reasoning_mode: None,
            assistant_use_separate_model: false,
            assistant_model: None,
            custom_providers: vec![
                custom_provider("a"),
                custom_provider("b"),
                custom_provider("c"),
            ],
        };

        assert_eq!(config.fallback_provider_after_removal("a"), "c");
    }

    #[test]
    fn split_reasoning_modes_fall_back_to_legacy_mode() {
        let config = LlmProviderConfig {
            active: "openai".to_string(),
            custom_base_url: None,
            custom_model: None,
            reasoning_mode: LlmReasoningMode::Light,
            polish_reasoning_mode: None,
            assistant_reasoning_mode: None,
            assistant_use_separate_model: false,
            assistant_model: None,
            custom_providers: Vec::new(),
        };

        assert_eq!(config.polish_reasoning_mode(), LlmReasoningMode::Light);
        assert_eq!(config.assistant_reasoning_mode(), LlmReasoningMode::Light);
    }
}

File: native-macos/Sources/LightWhisperNativeApp/Core/UserProfile.swift
import Foundation

public enum HotWordSource: String, Codable, Equatable, Sendable {
    case user
    case learned
}

public enum CorrectionSource: String, Codable, Equatable, Sendable {
    case ai
    case user
}

public struct HotWord: Codable, Equatable, Identifiable, Sendable {
    public var id: String { text }
    public var text: String
    public var weight: UInt8
    public var source: HotWordSource
    public var useCount: UInt32
    public var lastUsed: UInt64

    public init(
        text: String,
        weight: UInt8,
        source: HotWordSource,
        useCount: UInt32,
        lastUsed: UInt64
    ) {
        self.text = text
        self.weight = weight
        self.source = source
        self.useCount = useCount
        self.lastUsed = lastUsed
    }

    enum CodingKeys: String, CodingKey {
        case text
        case weight
        case source
        case useCount = "use_count"
        case lastUsed = "last_used"
    }
}

public struct CorrectionPattern: Codable, Equatable, Identifiable, Sendable {
    public var id: String { "\(original)->\(corrected)" }
    public var original: String
    public var corrected: String
    public var count: UInt32
    public var lastSeen: UInt64
    public var source: CorrectionSource

    enum CodingKeys: String, CodingKey {
        case original
        case corrected
        case count
        case lastSeen = "last_seen"
        case source
    }

    public init(
        original: String,
        corrected: String,
        count: UInt32,
        lastSeen: UInt64,
        source: CorrectionSource = .ai
    ) {
        self.original = original
        self.corrected = corrected
        self.count = count
        self.lastSeen = lastSeen
        self.source = source
    }

    public init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        original = try container.decode(String.self, forKey: .original)
        corrected = try container.decode(String.self, forKey: .corrected)
        count = try container.decode(UInt32.self, forKey: .count)
        lastSeen = try container.decode(UInt64.self, forKey: .lastSeen)
        source = try container.decodeIfPresent(CorrectionSource.self, forKey: .source) ?? .ai
    }
}

public struct VocabEntry: Codable, Equatable, Sendable {
    public var count: UInt32
    public var lastSeen: UInt64

    public init(count: UInt32, lastSeen: UInt64) {
        self.count = count
        self.lastSeen = lastSeen
    }

    enum CodingKeys: String, CodingKey {
        case count
        case lastSeen = "last_seen"
    }
}

public enum WebSearchProvider: String, Codable, CaseIterable, Equatable, Identifiable, Sendable {
    case modelNative = "model_native"
    case exa
    case tavily

    public var id: String { rawValue }
}

public struct WebSearchConfig: Codable, Equatable, Sendable {
    public var enabled: Bool
    public var provider: WebSearchProvider
    public var maxResults: UInt8

    enum CodingKeys: String, CodingKey {
        case enabled
        case provider
        case maxResults = "max_results"
    }

    public init(enabled: Bool = false, provider: WebSearchProvider = .modelNative, maxResults: UInt8 = 5) {
        self.enabled = enabled
        self.provider = provider
        self.maxResults = maxResults
    }

    public init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        enabled = try container.decodeIfPresent(Bool.self, forKey: .enabled) ?? false
        provider = try container.decodeIfPresent(WebSearchProvider.self, forKey: .provider) ?? .modelNative
        maxResults = try container.decodeIfPresent(UInt8.self, forKey: .maxResults) ?? 5
    }
}

public enum ApiFormat: String, Codable, CaseIterable, Equatable, Identifiable, Sendable {
    case openaiCompat = "openai_compat"
    case anthropic

    public var id: String { rawValue }
}

public enum OpenAIAuthMode: String, Codable, CaseIterable, Equatable, Identifiable, Sendable {
    case apiKey = "api_key"
    case oauth

    public var id: String { rawValue }
}

public enum LLMReasoningMode: String, Codable, CaseIterable, Equatable, Identifiable, Sendable {
    case providerDefault = "provider_default"
    case off
    case light
    case balanced
    case deep

    public var id: String { rawValue }
}

public struct CustomProvider: Codable, Equatable, Identifiable, Sendable {
    public var id: String
    public var name: String
    public var baseURL: String
    public var model: String
    public var apiFormat: ApiFormat

    enum CodingKeys: String, CodingKey {
        case id
        case name
        case baseURL = "base_url"
        case model
        case apiFormat = "api_format"
    }

    public init(
        id: String,
        name: String,
        baseURL: String,
        model: String,
        apiFormat: ApiFormat = .openaiCompat
    ) {
        self.id = id
        self.name = name
        self.baseURL = baseURL
        self.model = model
        self.apiFormat = apiFormat
    }

    public init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        id = try container.decode(String.self, forKey: .id)
        name = try container.decode(String.self, forKey: .name)
        baseURL = try container.decode(String.self, forKey: .baseURL)
        model = try container.decode(String.self, forKey: .model)
        apiFormat = try container.decodeIfPresent(ApiFormat.self, forKey: .apiFormat) ?? .openaiCompat
    }
}

public struct LLMProviderConfig: Codable, Equatable, Sendable {
    public var active: String
    public var customBaseURL: String?
    public var customModel: String?
    public var reasoningMode: LLMReasoningMode
    public var polishReasoningMode: LLMReasoningMode?
    public var assistantReasoningMode: LLMReasoningMode?
    public var assistantUseSeparateModel: Bool
    public var assistantModel: String?
    public var assistantProvider: String?
    public var customProviders: [CustomProvider]
    public var validationUseSeparateModel: Bool
    public var validationProvider: String?
    public var validationModel: String?
    public var openAIAuthMode: OpenAIAuthMode?
    public var openAIFastMode: Bool

    enum CodingKeys: String, CodingKey {
        case active
        case customBaseURL = "custom_base_url"
        case customModel = "custom_model"
        case reasoningMode = "reasoning_mode"
        case polishReasoningMode = "polish_reasoning_mode"
        case assistantReasoningMode = "assistant_reasoning_mode"
        case assistantUseSeparateModel = "assistant_use_separate_model"
        case assistantModel = "assistant_model"
        case assistantProvider = "assistant_provider"
        case customProviders = "custom_providers"
        case validationUseSeparateModel = "validation_use_separate_model"
        case validationProvider = "validation_provider"
        case validationModel = "validation_model"
        case openAIAuthMode = "openai_auth_mode"
        case openAIFastMode = "openai_fast_mode"
    }

    public static func defaultValue() -> LLMProviderConfig {
        LLMProviderConfig(
            active: "cerebras",
            customBaseURL: nil,
            customModel: nil,
            reasoningMode: .providerDefault,
            polishReasoningMode: nil,
            assistantReasoningMode: nil,
            assistantUseSeparateModel: false,
            assistantModel: nil,
            assistantProvider: nil,
            customProviders: [],
            validationUseSeparateModel: false,
            validationProvider: nil,
            validationModel: nil,
            openAIAuthMode: nil,
            openAIFastMode: false
        )
    }

    public init(
        active: String,
        customBaseURL: String?,
        customModel: String?,
        reasoningMode: LLMReasoningMode,
        polishReasoningMode: LLMReasoningMode?,
        assistantReasoningMode: LLMReasoningMode?,
        assistantUseSeparateModel: Bool,
        assistantModel: String?,
        assistantProvider: String?,
        customProviders: [CustomProvider],
        validationUseSeparateModel: Bool,
        validationProvider: String?,
        validationModel: String?,
        openAIAuthMode: OpenAIAuthMode?,
        openAIFastMode: Bool
    ) {
        self.active = active
        self.customBaseURL = customBaseURL
        self.customModel = customModel
        self.reasoningMode = reasoningMode
        self.polishReasoningMode = polishReasoningMode
        self.assistantReasoningMode = assistantReasoningMode
        self.assistantUseSeparateModel = assistantUseSeparateModel
        self.assistantModel = assistantModel
        self.assistantProvider = assistantProvider
        self.customProviders = customProviders
        self.validationUseSeparateModel = validationUseSeparateModel
        self.validationProvider = validationProvider
        self.validationModel = validationModel
        self.openAIAuthMode = openAIAuthMode
        self.openAIFastMode = openAIFastMode
    }

    public init() {
        self = Self.defaultValue()
    }

    public init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        let defaults = Self.defaultValue()
        active = try container.decodeIfPresent(String.self, forKey: .active) ?? defaults.active
        customBaseURL = try container.decodeIfPresent(String.self, forKey: .customBaseURL)
        customModel = try container.decodeIfPresent(String.self, forKey: .customModel)
        reasoningMode = try container.decodeIfPresent(LLMReasoningMode.self, forKey: .reasoningMode) ?? defaults.reasoningMode
        polishReasoningMode = try container.decodeIfPresent(LLMReasoningMode.self, forKey: .polishReasoningMode)
        assistantReasoningMode = try container.decodeIfPresent(LLMReasoningMode.self, forKey: .assistantReasoningMode)
        assistantUseSeparateModel = try container.decodeIfPresent(Bool.self, forKey: .assistantUseSeparateModel) ?? false
        assistantModel = try container.decodeIfPresent(String.self, forKey: .assistantModel)
        assistantProvider = try container.decodeIfPresent(String.self, forKey: .assistantProvider)
        customProviders = try container.decodeIfPresent([CustomProvider].self, forKey: .customProviders) ?? []
        validationUseSeparateModel = try container.decodeIfPresent(Bool.self, forKey: .validationUseSeparateModel) ?? false
        validationProvider = try container.decodeIfPresent(String.self, forKey: .validationProvider)
        validationModel = try container.decodeIfPresent(String.self, forKey: .validationModel)
        openAIAuthMode = try container.decodeIfPresent(OpenAIAuthMode.self, forKey: .openAIAuthMode)
        openAIFastMode = try container.decodeIfPresent(Bool.self, forKey: .openAIFastMode) ?? false
    }

    public func resolveActiveProvider() -> String {
        if Self.isBuiltinProvider(active) || customProviders.contains(where: { $0.id == active }) {
            return active
        }
        return customProviders.last?.id ?? "cerebras"
    }

    public func resolveAssistantProvider() -> String {
        guard assistantUseSeparateModel, let assistantProvider, !assistantProvider.isEmpty else {
            return resolveActiveProvider()
        }
        if Self.isBuiltinProvider(assistantProvider) || customProviders.contains(where: { $0.id == assistantProvider }) {
            return assistantProvider
        }
        return resolveActiveProvider()
    }

    public func resolveValidationProvider() -> String {
        guard validationUseSeparateModel, let validationProvider, !validationProvider.isEmpty else {
            return resolveActiveProvider()
        }
        if Self.isBuiltinProvider(validationProvider) || customProviders.contains(where: { $0.id == validationProvider }) {
            return validationProvider
        }
        return resolveActiveProvider()
    }

    public func resolvedPolishReasoningMode() -> LLMReasoningMode {
        polishReasoningMode ?? reasoningMode
    }

    public func resolvedAssistantReasoningMode() -> LLMReasoningMode {
        assistantReasoningMode ?? reasoningMode
    }

    public func resolvedAssistantModel() -> String? {
        guard assistantUseSeparateModel else {
            return nil
        }
        return assistantModel?.trimmingCharacters(in: .whitespacesAndNewlines).nilIfEmpty
    }

    public func resolvedValidationModel() -> String? {
        guard validationUseSeparateModel else {
            return nil
        }
        return validationModel?.trimmingCharacters(in: .whitespacesAndNewlines).nilIfEmpty
    }

    public static func isBuiltinProvider(_ value: String) -> Bool {
        ["cerebras", "openai", "deepseek", "siliconflow", "custom"].contains(value)
    }
}

public struct UserProfile: Codable, Equatable, Sendable {
    public var hotWords: [HotWord]
    public var correctionPatterns: [CorrectionPattern]
    public var vocabFrequency: [String: VocabEntry]
    public var totalTranscriptions: UInt64
    public var lastUpdated: UInt64
    public var llmProvider: LLMProviderConfig
    public var aiPolishEnabled: Bool
    public var dictationHotkey: String?
    public var translationTarget: String?
    public var translationHotkey: String?
    public var customPrompt: String?
    public var assistantHotkey: String?
    public var assistantSystemPrompt: String?
    public var assistantScreenContextEnabled: Bool
    public var aiPolishScreenContextEnabled: Bool
    public var blockedHotWords: [String]
    public var webSearch: WebSearchConfig
    public var correctionValidationEnabled: Bool
    public var lastCorrectionValidation: UInt64

    enum CodingKeys: String, CodingKey {
        case hotWords = "hot_words"
        case correctionPatterns = "correction_patterns"
        case vocabFrequency = "vocab_frequency"
        case totalTranscriptions = "total_transcriptions"
        case lastUpdated = "last_updated"
        case llmProvider = "llm_provider"
        case aiPolishEnabled = "ai_polish_enabled"
        case dictationHotkey = "dictation_hotkey"
        case translationTarget = "translation_target"
        case translationHotkey = "translation_hotkey"
        case customPrompt = "custom_prompt"
        case assistantHotkey = "assistant_hotkey"
        case assistantSystemPrompt = "assistant_system_prompt"
        case assistantScreenContextEnabled = "assistant_screen_context_enabled"
        case aiPolishScreenContextEnabled = "ai_polish_screen_context_enabled"
        case blockedHotWords = "blocked_hot_words"
        case webSearch = "web_search"
        case correctionValidationEnabled = "correction_validation_enabled"
        case lastCorrectionValidation = "last_correction_validation"
    }

    public static func defaultValue() -> UserProfile {
        UserProfile(
            hotWords: [],
            correctionPatterns: [],
            vocabFrequency: [:],
            totalTranscriptions: 0,
            lastUpdated: 0,
            llmProvider: LLMProviderConfig.defaultValue(),
            aiPolishEnabled: true,
            dictationHotkey: "f2",
            translationTarget: nil,
            translationHotkey: nil,
            customPrompt: nil,
            assistantHotkey: nil,
            assistantSystemPrompt: nil,
            assistantScreenContextEnabled: false,
            aiPolishScreenContextEnabled: false,
            blockedHotWords: [],
            webSearch: WebSearchConfig(),
            correctionValidationEnabled: false,
            lastCorrectionValidation: 0
        )
    }

    public init() {
        self = Self.defaultValue()
    }

    public init(
        hotWords: [HotWord],
        correctionPatterns: [CorrectionPattern],
        vocabFrequency: [String: VocabEntry],
        totalTranscriptions: UInt64,
        lastUpdated: UInt64,
        llmProvider: LLMProviderConfig,
        aiPolishEnabled: Bool,
        dictationHotkey: String?,
        translationTarget: String?,
        translationHotkey: String?,
        customPrompt: String?,
        assistantHotkey: String?,
        assistantSystemPrompt: String?,
        assistantScreenContextEnabled: Bool,
        aiPolishScreenContextEnabled: Bool,
        blockedHotWords: [String],
        webSearch: WebSearchConfig,
        correctionValidationEnabled: Bool,
        lastCorrectionValidation: UInt64
    ) {
        self.hotWords = hotWords
        self.correctionPatterns = correctionPatterns
        self.vocabFrequency = vocabFrequency
        self.totalTranscriptions = totalTranscriptions
        self.lastUpdated = lastUpdated
        self.llmProvider = llmProvider
        self.aiPolishEnabled = aiPolishEnabled
        self.dictationHotkey = dictationHotkey
        self.translationTarget = translationTarget
        self.translationHotkey = translationHotkey
        self.customPrompt = customPrompt
        self.assistantHotkey = assistantHotkey
        self.assistantSystemPrompt = assistantSystemPrompt
        self.assistantScreenContextEnabled = assistantScreenContextEnabled
        self.aiPolishScreenContextEnabled = aiPolishScreenContextEnabled
        self.blockedHotWords = blockedHotWords
        self.webSearch = webSearch
        self.correctionValidationEnabled = correctionValidationEnabled
        self.lastCorrectionValidation = lastCorrectionValidation
    }

    public init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        let defaults = Self.defaultValue()
        hotWords = try container.decodeIfPresent([HotWord].self, forKey: .hotWords) ?? defaults.hotWords
        correctionPatterns = try container.decodeIfPresent([CorrectionPattern].self, forKey: .correctionPatterns)
            ?? defaults.correctionPatterns
        vocabFrequency = try container.decodeIfPresent([String: VocabEntry].self, forKey: .vocabFrequency)
            ?? defaults.vocabFrequency
        totalTranscriptions = try container.decodeIfPresent(UInt64.self, forKey: .totalTranscriptions)
            ?? defaults.totalTranscriptions
        lastUpdated = try container.decodeIfPresent(UInt64.self, forKey: .lastUpdated) ?? defaults.lastUpdated
        llmProvider = try container.decodeIfPresent(LLMProviderConfig.self, forKey: .llmProvider) ?? defaults.llmProvider
        aiPolishEnabled = try container.decodeIfPresent(Bool.self, forKey: .aiPolishEnabled)
            ?? defaults.aiPolishEnabled
        dictationHotkey = try container.decodeIfPresent(String.self, forKey: .dictationHotkey)
            ?? defaults.dictationHotkey
        translationTarget = try container.decodeIfPresent(String.self, forKey: .translationTarget)
        translationHotkey = try container.decodeIfPresent(String.self, forKey: .translationHotkey)
        customPrompt = try container.decodeIfPresent(String.self, forKey: .customPrompt)
        assistantHotkey = try container.decodeIfPresent(String.self, forKey: .assistantHotkey)
        assistantSystemPrompt = try container.decodeIfPresent(String.self, forKey: .assistantSystemPrompt)
        assistantScreenContextEnabled = try container.decodeIfPresent(Bool.self, forKey: .assistantScreenContextEnabled)
            ?? defaults.assistantScreenContextEnabled
        aiPolishScreenContextEnabled = try container.decodeIfPresent(Bool.self, forKey: .aiPolishScreenContextEnabled)
            ?? defaults.aiPolishScreenContextEnabled
        blockedHotWords = try container.decodeIfPresent([String].self, forKey: .blockedHotWords)
            ?? defaults.blockedHotWords
        webSearch = try container.decodeIfPresent(WebSearchConfig.self, forKey: .webSearch) ?? defaults.webSearch
        correctionValidationEnabled = try container.decodeIfPresent(Bool.self, forKey: .correctionValidationEnabled)
            ?? defaults.correctionValidationEnabled
        lastCorrectionValidation = try container.decodeIfPresent(UInt64.self, forKey: .lastCorrectionValidation)
            ?? defaults.lastCorrectionValidation
    }

    public func hotWordTexts(limit: Int) -> [String] {
        hotWords
            .sorted {
                if $0.weight == $1.weight {
                    return $0.useCount > $1.useCount
                }
                return $0.weight > $1.weight
            }
            .prefix(limit)
            .map(\.text)
    }

    public func relevantCorrections(input: String, limit: Int) -> [CorrectionPattern] {
        correctionPatterns
            .filter { !$0.original.isEmpty && input.contains($0.original) }
            .sorted {
                if $0.source != $1.source {
                    return $0.source == .user
                }
                return $0.count > $1.count
            }
            .prefix(limit)
            .map { $0 }
    }
}

public enum UserProfileNormalizer {
    public static func normalize(_ profile: inout UserProfile) {
        migrateCustomProvider(&profile.llmProvider)
        migrateReasoningModes(&profile.llmProvider)
    }

    private static func migrateReasoningModes(_ config: inout LLMProviderConfig) {
        if config.polishReasoningMode == nil {
            config.polishReasoningMode = config.reasoningMode
        }
        if config.assistantReasoningMode == nil {
            config.assistantReasoningMode = config.reasoningMode
        }
    }

    private static func migrateCustomProvider(_ config: inout LLMProviderConfig) {
        guard config.active == "custom", config.customProviders.isEmpty else {
            return
        }

        let baseURL = config.customBaseURL?.trimmingCharacters(in: .whitespacesAndNewlines) ?? ""
        let model = config.customModel?.trimmingCharacters(in: .whitespacesAndNewlines) ?? ""
        guard !baseURL.isEmpty || !model.isEmpty else {
            return
        }

        config.customProviders = [
            CustomProvider(
                id: "custom_migrated",
                name: "Custom Compatible",
                baseURL: baseURL,
                model: model,
                apiFormat: .openaiCompat
            ),
        ]
        config.active = "custom_migrated"
        config.customBaseURL = nil
        config.customModel = nil
    }
}

private extension String {
    var nilIfEmpty: String? {
        isEmpty ? nil : self
    }
}
